ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -1008,6 +1008,10 @@ private void doPhase1GetAllAggregations(ASTNode expressionTree, QB qb,
   }
 
   private void transformWithinGroup(ASTNode expressionTree, Tree withinGroupNode) throws SemanticException {
+    if (HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_CBO_ENABLED) && !ctx.isCboPlanReady()) {
+      return;
+    }
+
```

Comment:
Why do we to skip this transformation conditionally and why it is guarded by two variables?

I can understand that we may want to do some special processing to transform `WITHIN GROUP` when CBO is disabled but when it is enabled why do we need to add `isCboPlanReady`?

I suppose it has something to do with the comment below but I don't fully understand it yet.

> The UDF implementation haven't change so when the second compiling phase is done after CBO or CBO is off the transformation is still applied.

The typical compilation path is the following: `AST -> RelNode -> Optimized AST -> Operator` 
Can you highlight at which point(s) we call the `transformWithinGroup` in this pipeline?

What happens if CBO failed? Is it safe to rely on `isCboPlanReady`?

Reply:
If CBO on `transformWithinGroup` called on the `Optimized AST` only.
If CBO off is is called on the initial `AST`
If CBO fails we fall back to non-CBO path and the transformation is called on initial `AST`.

I had to introduce `isCboPlanReady` because `isCboSucceeded` set to `true` only after phase1 analysis completed on `Optimized AST`.

Reply:
> Is this change gonna affect EXPLAIN CBO output?

Yes, Sort keys are shown as collation of the aggregate function call

Reply:
If we don't have EXPLAIN CBO with collation and sort keys it would be nice to add one/two test cases since someone may break this feature in the future without realizing. Also if a rule drops this information for some reason then it will probably lead to wrong results so it would be good to have a few basic tests.

I see that in some other places in SemanticAnalyzer we are making use of `isCBOExecuted()` method to do things; is it applicable here?

Reply:
I added test for explain cbo with collation.

`isCBOExecuted` shows only whether CBO is enabled. I think it has a confusing method name because it only gets the value of the field `runCBO` in `CalcitePlanner` and returns `false` in `SemanticAnalyzer`.
Anyway, I replaced querying the config HIVE_CBO_ENABLED with it.

Reply:
I was hoping that maybe `isCBOExecuted` was enough to replace also the changes in the context but maybe its not the case.